File: backend/app/models/sleep_log.py
import datetime

from sqlalchemy import DateTime, ForeignKey, Index, Integer, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class SleepLog(Base):
    __tablename__ = "sleep_logs"
    __table_args__ = (Index("idx_sleep_logs_baby_started", "baby_id", "started_at"),)

    id: Mapped[str] = mapped_column(String, primary_key=True)
    family_id: Mapped[str] = mapped_column(
        String, ForeignKey("families.id"), nullable=False
    )
    baby_id: Mapped[str] = mapped_column(
        String, ForeignKey("babies.id"), nullable=False
    )
    started_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    ended_at: Mapped[datetime.datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    note: Mapped[str | None] = mapped_column(String, nullable=True)
    client_id: Mapped[str | None] = mapped_column(String, nullable=True)
    schema_version: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    created_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    updated_at: Mapped[datetime.datetime] = mapped_column(
        DateTime(timezone=True), nullable=False
    )
    deleted_at: Mapped[datetime.datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

File: frontend/app/src/main/java/com/nurtlina/app/domain/model/SleepLog.kt
package com.nurtlina.app.domain.model

import java.time.Instant

data class SleepLog(
    val id: String,
    val babyId: String,
    val startedAt: Instant,
    val endedAt: Instant?,
    val note: String?,
    val createdAt: Instant,
    val updatedAt: Instant,
) {
    val isActive: Boolean get() = endedAt == null

    fun durationMillis(): Long? = endedAt?.let { it.toEpochMilli() - startedAt.toEpochMilli() }
}
